### PR TITLE
allow passing in extra headers to requests. 

### PR DESCRIPTION
**Pull request by williamzhuk**
Test will use az etags logic in the header

### CODE REVIEW

File: blobfile/_ops_test.py [hunk 1]
```diff
@@ -1473,6 +1473,28 @@ def test_azure_maybe_update_md5(ctx):
...
 
+@pytest.mark.parametrize("ctx", [_get_temp_as_path])
+def test_azure_etags(ctx):
+    contents = b"bark!"
+    alternative_contents = b"ruff"
+
+    with ctx() as path:
+        bf.BlobFile(path, "wb").write(contents)
+        etag = bf.stat(path).version
+        with bf.create_context(additional_http_headers={"If-Match": etag}) as ctx:
```
Comment (christopher-hesse) [inline blobfile/_ops_test.py:1484]: I'm not certain how this test code worked even in a standalone script

**Comment by christopher-hesse** [comment]
I appreciate the PR and test, but I couldn't get the test to pass in https://github.com/christopher-hesse/blobfile/pull/208

I think I could merge this if the test passed there, but I'm curious about a bigger issue here, which is that we're overriding all requests in the context.  If this is only used for writing files, would it make more sense to add this etag as part of the `bf.BlobFile` constructor? e.g. `bf.BlobFile(path, "rb", require_version=version)` or something along those lines.

File: blobfile/_context.py [hunk 2]
```diff
@@ -115,7 +116,7 @@ def copy(
                 copy_fn = _parallel_download
 
             if _is_local_path(src) and _is_azure_path(dst):
-                copy_fn = azure.parallel_upload
+                copy_fn = partial(azure.parallel_upload, version=version)
```
Comment (williamzhuk) [inline blobfile/_context.py:119]: not sure if this is the cleanest way of doing this, but will cause fail-fast trying to pass versions into gcp or local fns.

Comment (christopher-hesse) [reply blobfile/_context.py:119]: I don't follow, how does this cause a failure when using GCP?

**Comment by christopher-hesse** [comment]
Closing in favor of https://github.com/christopher-hesse/blobfile/pull/209/files, I can merge this after the above issue is resolved

File: blobfile/_ops_test.py [hunk 3]
```diff
@@ -1473,6 +1473,38 @@ def test_azure_maybe_update_md5(ctx):
         )
 
 
+@pytest.mark.parametrize("ctx", [_get_temp_as_path])
+def test_azure_etags(ctx):
+    contents = b"bark!"
```
Comment (christopher-hesse) [inline blobfile/_ops_test.py:1478]: Won't this scare all the cat tests?

File: blobfile/_context.py (hunk 2 above)
Comment (christopher-hesse) [reply blobfile/_context.py:119]: I think the remote copy case was missed as well, going to fix some things up on the other PR, might have to merge tomorrow.

**Comment by christopher-hesse** [comment]
Should be fixed now

**Pull request closed without merging**